app/Jobs/CreateQueryserviceBatchesJob.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+namespace App\Jobs;
+
+use App\EventPageUpdate;
+use App\QsBatch;
+use App\QsCheckpoint;
+use Illuminate\Database\Eloquent\Collection;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Config;
+
+class CreateQueryserviceBatchesJob extends Job
+{
+    private const NAMESPACE_ITEM = 120;
+    private const NAMESPACE_PROPERTY = 122;
+    private const NAMESPACE_LEXEME = 146;
+
+    private int $entityLimit;
+
+    public function __construct()
+    {
+        $this->entityLimit = Config::get('wbstack.qs_batch_entity_limit');
+    }
+
+    public function handle(): void
+    {
+        DB::transaction(function () {
+            $lastCheckpoint = QsCheckpoint::get();
+            $latestEventId = $lastCheckpoint;
+            $newEntities = $this->getNewEntities($lastCheckpoint, $latestEventId);
+
+            foreach ($newEntities as $wikiId => $entityIdsFromEvents) {
+                $ok = $this->tryToAppendEntitesToExistingBatches($entityIdsFromEvents, $wikiId);
+                if ($ok) {
+                    continue;
+                }
+                $this->createNewBatches($entityIdsFromEvents, $wikiId);
+            }
+
+            QsCheckpoint::set($latestEventId);
+        });
+    }
+
+    private function getNewEntities($lastCheckpoint, &$latestEventId): array
+    {
+        $newEntitiesFromEvents = [];
+
+        $events = EventPageUpdate::where(
+            'id', '>', $lastCheckpoint
+        )->get();
+
+        foreach ($events as $event) {
+            if (
+                $event->namespace == self::NAMESPACE_ITEM ||
+                $event->namespace == self::NAMESPACE_PROPERTY ||
+                $event->namespace == self::NAMESPACE_LEXEME
+            ) {
+                $newEntitiesFromEvents[$event->wiki_id][] = $event->title;
+            }
+            if ($event->id > $latestEventId) {
+                $latestEventId = $event->id;
+            }
+        }
+
+        return $newEntitiesFromEvents;
+    }
+
+    private function tryToAppendEntitesToExistingBatches(array $entityIdsFromEvents, int $wikiId): bool
+    {
+        $notDoneBatches = QsBatch::where([
```

Comment:
I moved this query a level down which means:
- we now do one query per wiki (i.e. more SQL queries being run)
- less manual filtering of wikis that don't apply

In the end it's probably a bit slower, but considering the query is not really complicated I think the fact that it lets us remove yet another if condition in a for loop is worth the performance hit. Let me know if you think otherwise.